Implement Cobuild integration and Solana support

XuJiandong opened the pull request:
For the detail of Solana support, please see https://github.com/XuJiandong/omnilock/pull/48

Comments for Cobuild reviewing:

* The main implementation is in cobuild.h.
* Otx(open transaction) implementation is included.
* The words "tested by" in cobuild.c means the branch is covered by corresponding test case.
* The lazy reader is implemented in cobuild.c. No matter how large the transaction is, the total memory consumption is fixed(~2K for global and ~4K for stack). We have test case for it.
* The precomputed table for secp256k1 is removed. It can save a lot of memory (~1M).
* molecule data structure is verified with lazy reader(See verfiy_WitnessLayout)
* There is [message calculation flow](https://github.com/XuJiandong/omnilock/blob/f7e65908c1dbb60ea373a38c0142ddaf5bbb1265/c/cobuild.h#L489-L490) in omnilock which doesn't exist in cobuild spec. One extra byte at start of seal is for further use.

Commit:
Add missing auth id(0x02~0x05)

migrate PR from: https://github.com/nervosnetwork/ckb-production-scripts/pull/75

including
* eos
* tron
* bitcoin(Support UniSat and OKX wallet)
* dogecoin

Special feature for btc/etc, now they can display meaningful messages.

* BTC(UniSat/OKX)
You're signing:
CKB (Bitcoin Layer-2) transaction: 0x{sighash_all in hex}

* ETH(Metamask)
You're signing:
CKB transaction: 0x{sighash_all in hex}

Commit:
Cobuild support

* The main implementation is in cobuild.c.
* The lazy reader is implemented in cobuild.c.

Other changes:
* Add test cases for cobuild
* Update ckb-* to 0.113.0 and rust-toolchain
* Update ckb-c-stdlib
* Add test vectors

xxuejie commented on c/omni_lock.c:
```diff
@@ -396,8 +398,22 @@ int main() {
+  } else {
+    printf("cobuild enabled");
+    g_cobuild_enabled = true;
+    err = parse_witness_lock(&witness_lock, &seal);
```
If we want to move forward with this change, I would suggest that a prefix byte should be added at least here to seal structure. The prefix byte can then be used to control how the signing message is generated(much like the algorithm id in script args controlling signing algorithm to use). Right now the prefix byte might accept only one value that represents the blake2b based message calculation flow. But in the future, this very prefix byte, can then be used to introduce more variations on signing message calculation, such as a structured text based signing message

XuJiandong replied:
So we're going to implement it below, without molecule encoding:
```text
<message calculation flow, 1 byte> <seal part>
```

XuJiandong committed:
Check type scripts from message existing in tx (#14)

* support message calculation flow in seal
* Add testcase

mohanson committed:
Use CKB 2nd hardfork (#16)

* Change to 1st hardfork

XuJiandong committed:
Support cobuild OTX (#18)

Cobuild otx implementation and test cases.

Other major changes:
* Minimize secp256k1 precomputed table
* Support llvm toolchain

Other minor changes:
* Update gcc toolchain
* Update ckb-c-stdlib
* Add rust format rules

---------

Co-authored-by: Mohanson <mohanson@outlook.com>
Co-authored-by: joii2020 <87224197+joii2020@users.noreply.github.com>

xxuejie commented on Makefile:
```diff
@@ -3,12 +3,15 @@ CC := $(TARGET)-gcc
 OBJCOPY := $(TARGET)-objcopy
 CFLAGS := -fPIC -O3 -fno-builtin-printf -fno-builtin-memcmp -nostdinc -nostdlib -nostartfiles -fvisibility=hidden -fdata-sections -ffunction-sections -I deps/secp256k1/src -I deps/secp256k1 -I deps/ckb-c-std-lib -I deps/ckb-c-std-lib/libc -I deps/ckb-c-std-lib/molecule -I c -I build -Wall -Werror -Wno-nonnull -Wno-nonnull-compare -Wno-unused-function -g
-LDFLAGS := -Wl,-static -fdata-sections -ffunction-sections -Wl,--gc-sections
+LDFLAGS := -nostdlib -nostartfiles -fno-builtin -Wl,-static -Wl,--gc-sections
```
I would personally only keep a single makefile, either we use clang, or due to certain issues, we stick with clang. It brings confusion if we provide 2 here.

XuJiandong replied:
we will use clang version when the official docker image (crytape) is ready.

xxuejie replied:
This is okay, but I do suggest that makefile for clang can be removed from this PR. In a later PR, we can switch the build process from gcc to clang.

The point is, I don't think 2 makefiles are needed at the same time.

xxuejie commented on c/dump_secp256k1_data_20210801.c:
```diff
@@ -34,22 +34,23 @@ int main(int argc, char* argv[]) {
-  for (int i = 0; i < 32; i++) {
-    fprintf(fp, "%u", hash[i]);
-    if (i != 31) {
+  fprintf(fp, "static uint8_t ckb_secp256k1_data[] = {\n  ");
```
I was slightly confused here. We have 2 source files for dumping secp data: `dump_secp256k1_data.c` and `dump_secp256k1_data_20210801.c`. If we are removing one, maybe we should remove that `dump_secp256k1_data_20210801.c` one?

XuJiandong replied:
The "20210801" is the tag for files and folders only. It means that we use a secp256k1 version which is labeled at "2021/08/01". The variables are not following this conversions.

In this way, we can multiple secp256k1 and ckb-c-stdlib in same project without conflicts. So we will remove "dump_secp256k1_data.c".

xxuejie replied:
This solution makes sense for `ckb-production-scripts` repos, where we might have multiple contracts developed at different time using different dependencies. But now we are already keeping omnilock contract in a single repository. I don't see the need to use the same convention. In addition, the exact git commit of submodule already provides a `version` for the dependency used. Personally I don't see the value of timestamp here.

xxuejie commented on c/omni_lock.c:
```diff
@@ -1,18 +1,15 @@
-// #define CKB_C_STDLIB_PRINTF
-
+// in secp256k1_ctz64_var: we don't have __builtin_ctzl in gcc for RISC-V
+#define __builtin_ctzl secp256k1_ctz64_var_debruijn
```
Latest clang / gcc, with B extention enabled, both have `__builtin_ctzl` implemented using a single RISC-V instruction. I don't think there is the need to use secp256k1_ctz64_var_debruijn which is a slower software implementation

xxuejie commented on c/cobuild.c:
```diff
@@ -0,0 +1,1043 @@
+      break;
+    }
+    mol2_union_t uni = mol2_union_unpack(&cursor);
+    OtxType otx = make_Otx(&uni.cursor);
```
One big thing that is troubling me, is that the `make_*` APIs provided by molecule-c2, does no verification at all. Yes it does basic checking for the part that you **accessed**, but in general, those `make_*` calls do not verify that a witness(or another piece of data) fully confronts the molecule schema we were defining.

As cobuild is build surrounding the idea of playing with molecule data verification scheme, skipping full verification here makes me worried, that we might see unexpected piece of data achieving unwanted results as usage grows.

If I were working on omnilock, I would refocus, and enforce full verification on molecule-c2

xxuejie commented on c/cobuild.c:
```diff
@@ -0,0 +1,1043 @@
+  if (id_len < sizeof(id)) {
+    return ERROR_GENERAL;
+  }
+  if (id == WitnessLayoutSighashAll || id == WitnessLayoutSighashAllOnly ||
```
Are we hitting limitations somewhere? Why not just build a full WitnessLayout structure?

XuJiandong replied:
This is due to the limitation of molecule-c2. If the data structure is invalid, it may quit(panic) early. 
Note, it's hard to make it panic free to parse an invalid data. e.g. overflow.

We need to figure out what  the data structure is before doing verification.

xxuejie replied:
I was confused: if a witness data structure is invalid, the smart contract should abort anyway. What's the problem of quiting / panicing / aborting here?

XuJiandong replied:
Actually the "get_witness_layout" should be renamed to "detect_witness_layout" here. It's possible that the data structure is WitnessArgs or invalid.

xxuejie replied:
Let's try this again:

* Make a WitnessLayout data structure from a cursor.
* Run a verifier to ensure the correctness of the WitnessLayout data structure. A verifier here should be panic free, regardless of what data is passed here.
* If the verifier passes, you have a correct data structure, no panic will happen afterwards
* If the verifier fails, you try the next one

XuJiandong replied:
Currently, molecule-c2 is not panic free.

xxuejie replied:
I'm not talking about molecule-c2, which still does not have a verifier. I'm talking about potential verifier needed here. molecule-c2's most current accessor code can stay as it is with the potential of panicing at times, as long as we can provide verifier code that is panic free(and should be panic free, we have already done similar thing in Rust)

XuJiandong replied:
Panic free is not very easy. Every operator (+, -) should be checked to avoid overflowing.

xxuejie replied:
This is not about easy or hard. You are merely talking about efforts and paying attention to the code when you write it. If we don't have a solution to a thing, I would call it hard. Here it is just about efforts.

I can experiment on it if you guys find it hard to do.

xxuejie reviewed (commented):
I haven't finished checking the correctness part of the code.

But the more I look at this code, the more I think we failed to deliver a proper set of API in molecule-c2, and the accompanying Rust API at https://github.com/nervosnetwork/molecule/pull/77.

Personally, I would picture 2 clearly defined, distinct eras of APIs where CKB smart contracts use:

* Initially, we had a bunch of APIs, like load_input, load_witness, load_cell, load_cell_by_field, etc. And you are expected to be experienced enough to mix the use of them to achieve validation
* With lazy loader, we should be able to reduce the effort to only a selected few APIs: load_script for loading current script, and load_transaction for accessing anything related to the enclosing transaction.

In the second era, I would expect a smart contract only needs to build a single data source containing the entire transaction, and in 99% of the case, CKB transaction will very small, meaning a single syscall can be used by the data source to load the whole transaction into a memory buffer owned by the data source. Later the smart contract only needs to deal with a single cursor built from the data source containing the full transaction. The smart contract can then slice the cursor into smaller cursors, like one cursor for an input cell, and another cursor for an output cell lock, and yet another cursor for a witness. A cursor representing a witness can be further reassembly into a cursor for a WitnessLayout structure, or a WitnessArgs structure for legacy layout.

To me, that would bring us a much cleaner and a much performant design.

However, in current omnilock, we are still mixing numerous APIs together, littered with different patterns. `get_witness_layout` and `ckb_fetch_otx_start` are 2 examples here. We are manually coding against many hidden jargons, making the code harder to read nor maintain. Combining with the fact that molecule-c2 does no verification on the data structure, I must say I wouldn't personally agree with the way this lock is organized now.

Yes there is a chance the lock script might still be correct mostly and secure, but to me this lock is moving into a uncertain direction, and I'm afraid it's gonna bite us one day.

So if I were doing this, I would go back to molecule-c2, if the API is not good enough for usage, we should alter molecule-c2 with better APIs. After that, omnilock should only need to construct a single cursor containing the full CKB transaction, and continue to validate from there.

XuJiandong commented:
> In the second era, I would expect a smart contract only needs to build a single data source containing the entire transaction

A single data source is not working for following data structure:
1. cell_deps
2. header
3. inputs

xxuejie commented:
> > In the second era, I would expect a smart contract only needs to build a single data source containing the entire transaction
> 
> A single data source is not working for following data structure:
> 
> 1. cell_deps
> 2. header
> 3. inputs

Correct, those syscalls are just like `load_script`, which is outside of a single transaction, and should be loaded via designated syscalls.

But that does not void the reasoning here: there are simply too many code pieces in the implementation now to load a part of the data which indeed belong to the enclosing transaction, causing enough confusions and chaos.

joii2020 committed:
Update according to reviews (#38)

* Remove all "20210801". Include: submodules(ckb-c-stdlib, secp256k1), secp256k1_data
* Remove clang makefile
* Use compiled __builtin_ctzl

XuJiandong committed:
Fix relative since issue (#39)

* Fix check_since
* Add test cases for since

xxuejie committed:
Cobuild refactor (#40)

* Refactor to leverage molecule-c2 as much as possible

xxuejie commented:
This can of course be a latter task, but I believe we can move `cobuild.c` / `cobuild.h` (ideally I would personally just keep a `cobuild.h` file containing both declaration and implementation for the ease of integration, but that a question of preference) to `ckb-c-stdlib` for reuse in other contracts.

joii2020 committed:
Validate witnesses (#42)

Apply verify_WitnessLayout and verify_WitnessArgs on molecule data.

---------

Co-authored-by: xjd <xjd@cryptape.com>

XuJiandong committed:
Validate actions to lock script hashes (#43)

jjyr commented on c/basic.mol:
```diff
@@ -0,0 +1,81 @@
+vector ScriptInfoVec <ScriptInfo>;
+
+table ResolvedInputs {
+		outputs: CellOutputVec,
```
There are several inconsist indent are used in this file, pls fix them

joii2020 committed:
Add testcases (#44)

Add test cases:
* The witnesses is none
* The output data is big enough(700K)
* Add a 700K witness
* Args is big enough (700K)

joii2020 committed:
Testcase Otx support header_dep (#45)

* Test case Otx support header_dep

joii2020 committed:
Add test vectors for cobuild (#46)

* Add test vectors for cobuild
* Remove deps secp256k1_data on testcase

homura commented:
Hi there, it's Lumos. We're planning to integrate with the CoBuild protocol through Omnilock and would like to ask a few questions before proceeding:

- Is there any documentation or example that demonstrates how Omnilock supports the CoBuild protocol
- Is the PR stable and unlikely to undergo major changes that could affect the integration process
- Do you have any other suggestions for SDKs that would be useful in integrating with this PR

XuJiandong commented:
> Hi there, it's Lumos. We're planning to integrate with the CoBuild protocol through Omnilock and would like to ask a few questions before proceeding:
> 
> * Is there any documentation or example that demonstrates how Omnilock supports the CoBuild protocol
> * Is the PR stable and unlikely to undergo major changes that could affect the integration process
> * Do you have any other suggestions for SDKs that would be useful in integrating with this PR

issue 2: yes, it's stable.
issue 3: We have test vectors in this PR. It will be useful.

EthanYuan commented:
Hello, CKB SDK Rust is also planning to support CoBuild , it would be very helpful to have a synchronized PR for updating [0042-omnilock.md](https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0042-omnilock/0042-omnilock.md).

doitian commented on c/top_level.mol:
```diff
@@ -0,0 +1,8 @@
+import basic;
```
Contract usually only needs `WitnessLayout`, I would recommend spliting the schema into two files:

- `witness_layout.mol`: Contains `WitnessLayout` and its dependencies
- `building_packet.mol`: This is for offchain, it imports `witness_layout` and includes the definition for `BuildingPacket` as well as other dependencies.

doitian commented on c/omni_lock_time_lock.h:
```diff
@@ -1,73 +1,18 @@
  constraint, and the value of since must greater or equals than the since
- contstaint */
+ constraint */
 int check_since(uint64_t since) {
```
I found the function hard to reason. Maybe renaming the paramemter can make it easier.

```suggestion
int check_since(uint64_t minimal_input_since_constraint) {
```

XuJiandong replied:
I'm feeling not good about a so long  local variable. Can we have it shorter like "input_since"?

doitian replied:
How about `min_since`

homura commented:
I have a question regarding the script built by `make all-via-docker`

I remember that the `printf` instructions used to be ignored or erased by the compiler. However, the compiled binary from this PR still contains the `printf` instructions. I am not sure if this is the expected behavior

I received the following output from the ckb-debugger, where the log `No otx detected` may have been generated from [this line](https://github.com/XuJiandong/omnilock/blob/91efc8ac0018e391e6d29c1cd41f69b16fa5d89d/c/cobuild.c#L724):

```
No otx detected
message(len=12): 0c0000000800000004000000
seal(len=86): 00550000001000000055000000550000004100000016c86c5dfed70f32cf0b62e31517b7b376de7453a4e1637d7d4300f374a6ef7061183fef1f116073e2b235496ca96a2c2eb8b9c03e76cb0b5b88b427bfe0953a01
ckb_generate_smh total hashed 200 bytes
smh(len=32): 55009e27578c58aaf1255a713d525770f0992a9c2e09604fba4221c44fe2dded
cobuild_activated = 1
Run result: 0
Total cycles consumed: 1368092(1.3M)
Transfer cycles: 30437(29.7K), running cycles: 1337655(1.3M)
```

XuJiandong commented:
> I have a question regarding the script built by `make all-via-docker`
> 
> I remember that the `printf` instructions used to be ignored or erased by the compiler. However, the compiled binary from this PR still contains the `printf` instructions. I am not sure if this is the expected behavior.
> 
> I received the following output from the ckb-debugger, where the log `No otx detected` may have been generated from [this line](https://github.com/XuJiandong/omnilock/blob/91efc8ac0018e391e6d29c1cd41f69b16fa5d89d/c/cobuild.c#L724):
> 
> ```
> No otx detected
> message(len=12): 0c0000000800000004000000
> seal(len=86): 00550000001000000055000000550000004100000016c86c5dfed70f32cf0b62e31517b7b376de7453a4e1637d7d4300f374a6ef7061183fef1f116073e2b235496ca96a2c2eb8b9c03e76cb0b5b88b427bfe0953a01
> ckb_generate_smh total hashed 200 bytes
> smh(len=32): 55009e27578c58aaf1255a713d525770f0992a9c2e09604fba4221c44fe2dded
> cobuild_activated = 1
> Run result: 0
> Total cycles consumed: 1368092(1.3M)
> Transfer cycles: 30437(29.7K), running cycles: 1337655(1.3M)
> ```
> 
> another possibility is that the `printf` instruction is not erased, but that the debugger I used before may not set an [output printer](https://github.com/nervosnetwork/ckb-standalone-debugger/blob/1b35d718b4e5a7e01790cc17db0f67ef93b2cec1/ckb-debugger/src/main.rs#L297-L302), whereas the current debugger has one

Will be removed when it's ready. It's very important for troubleshooting now.

homura commented:
> It is expected behavior. Will be removed when it's ready. It's very important for troubleshooting now.

I see, it does helpful for debugging

XuJiandong committed:
Refactor cobuild into a standalone library (#47)

* refactor cobuild into a standalone library
* fix simulator bug on witness
* cleanup comments
* update clang-format docker

homura commented:
I noticed that during the scanning process, the contract checks for the presence of a [WitnessLayout in witnesses](https://github.com/XuJiandong/omnilock/blob/fd6674d8d4a257cd378d0cc2fef60d362d85878a/c/cobuild.h#L714-L720) to determine if CoBuild mode is activated. However, the current implementation seems incomplete in scenarios where an NFT can be minted while unlocking a long-term Nervos DAO. Such transactions may look like this:

```yaml
inputs:
  - input#0:
      lock: omnilock
      type: dao
  - input#1 # mint lock cell
      lock: nft-mint-lock # to check if a Nervos DAO has been locked for a particular period
outputs:
  - output#0:
      lock: omnilock
      type: dao-withdraw
  - output#1 # mint lock cell
  - output#2 # NFT cell
witnesses:
  - witness#0: WitnessArgs # dao deposit header witness
  - witness#1: WitnessLayout.SighashAll # with mint NFT action
```

XuJiandong committed:
Solana (phantom wallet) support (#48)

* Solana (phantom wallet) draft support
* Check consumed cycles
* Optimize get_witness_layout
* Add macro DISABLE_MESSAGE_CALCULATION_FLOW

XuJiandong committed:
Disable log (#49)

Add checking to generated binary

XuJiandong commented:
Pudge testnet deployment information:
| parameter   | value                                                                |
| ----------- | -------------------------------------------------------------------- |
| `code_hash` | 0x533cbfca1aba7b6286d07c23292f80b89f690786bc1cc07c96f0b6643849fc5d   |
| `hash_type` | `type`                                                               |
| `tx_hash`   |0xbf5be776ebc34bf681606a197116eaf5f4b23ada55cc5fc32f7abbda081d0918  |
| `index`     | `0x0`                                                                |
| `dep_type`  | `code`                                                               |

XuJiandong committed:
Add exec test case (#50)

* Add exec test case
